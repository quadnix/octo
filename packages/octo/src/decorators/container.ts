import type { Constructable } from '../app.type.js';

type Factory<T> = { create: (...args: unknown[]) => Promise<T> };
type FactoryValue<T> = Factory<T> | [Promise<Factory<T>>, (factory: Factory<T>) => void];

/**
 * The Container class is a box to hold all registered factories.
 *
 * In Octo, we minimize the use of the `new` keyword and get instance of a class using its factory.
 * The factory has full control over how the instance is created.
 *
 * By using the Container and Factory concepts, it is possible to override internal class definitions at runtime.
 * It is an incredibly powerful tool to customize implementation.
 */
export class Container {
  private static FACTORY_TIMEOUT_IN_MS = 5000;

  private static readonly factories: {
    [key: string]: {
      default: boolean;
      factory: FactoryValue<any>;
      metadata: { [key: string]: string };
    }[];
  } = {};

  /**
   * `Container.get()` allows to get an instance of a class using its factory.
   *
   * @param type The type or name of the class to get.
   * @param options Allows selection of a specific factory to use to get the instance.
   * - The `args` option supplies arguments to the factory.
   * - The `metadata` option identifies the factory.
   * @returns The instance of the class.
   */
  static async get<T>(
    type: Constructable<T> | string,
    options?: {
      args?: unknown[];
      metadata?: { [key: string]: string };
    },
  ): Promise<T> {
    const args = options?.args || [];
    const metadata = options?.metadata;
    const name = typeof type === 'string' ? type : type.name;

    if (!(name in this.factories)) {
      this.factories[name] = [];
    }

    let factoryContainer: (typeof Container.factories)[keyof typeof Container.factories][0] | undefined;
    if (!metadata) {
      factoryContainer = this.factories[name].find((f) => f.default);
    } else {
      const filters: { key: string; value: string }[] = [];
      for (const key in metadata) {
        filters.push({ key, value: metadata[key] });
      }
      factoryContainer = this.factories[name].find((f) => filters.every((c) => f.metadata[c.key] === c.value));
    }

    if (factoryContainer?.factory) {
      if (!Array.isArray(factoryContainer.factory)) {
        return (factoryContainer.factory as Factory<T>).create(...args);
      } else {
        const factory = await (factoryContainer.factory[0] as Promise<Factory<T>>);
        return factory.create(...args);
      }
    }

    const newFactoryContainer = {
      default: true,
      metadata,
    } as (typeof Container.factories)[keyof typeof Container.factories][0];
    this.factories[name].push(newFactoryContainer);

    let promiseResolver;
    let promiseTimeout;
    const promise = new Promise<Factory<T>>((resolve, reject) => {
      promiseTimeout = setTimeout(() => {
        reject(new Error(`Timed out waiting for factory ${name} to resolve!`));
      }, this.FACTORY_TIMEOUT_IN_MS);
      promiseResolver = resolve;
    });
    newFactoryContainer.factory = [promise, promiseResolver];

    const factory = await promise;
    if (promiseTimeout) {
      clearTimeout(promiseTimeout);
    }
    return factory.create(...args);
  }

  /**
   * `Container.registerFactory()` allows to register a factory for a class.
   *
   * @param type The type or name of the class for which the factory is registered.
   * @param factory The factory class being registered.
   * @param options Distinguishes between different factories of the same class.
   * - The `metadata` attaches custom metadata to the factory.
   */
  static registerFactory<T>(
    type: Constructable<T> | string,
    factory: Factory<T>,
    options?: {
      metadata?: { [key: string]: string };
    },
  ): void {
    const metadata = options?.metadata || {};
    const name = typeof type === 'string' ? type : type.name;

    if (!(name in this.factories)) {
      this.factories[name] = [];
    }

    const filters: { key: string; value: string }[] = [];
    for (const key in metadata) {
      filters.push({ key, value: metadata[key] });
    }
    const factoryContainer = this.factories[name].find((f) => filters.every((c) => f.metadata[c.key] === c.value));

    if (factoryContainer?.factory) {
      // If factory is not a promise set by get() above, it has already been registered.
      if (!Array.isArray(factoryContainer.factory)) {
        return;
      }

      factoryContainer.metadata = metadata;
      const resolve = factoryContainer.factory[1] as (factory: Factory<T>) => void;
      resolve(factory);
      factoryContainer.factory = factory;
      this.setDefault(type, factory);

      return;
    }

    this.factories[name].push({ default: false, factory, metadata });
    this.setDefault(type, factory);
  }

  /**
   * `Container.reset()` clears all registered factories and empties the container. This is mostly used in testing.
   */
  static reset(): void {
    this.FACTORY_TIMEOUT_IN_MS = 5000;

    for (const name in this.factories) {
      delete this.factories[name];
    }
  }

  /**
   * `Container.setDefault()` sets a default factory for the class.
   *
   * @example
   * ```ts
   * // Register a factory.
   * Container.registerFactory(MyClass, MyClassFactory, { metadata: { key: 'value' } });
   *
   * // Without default.
   * Container.get(MyClass, { metadata: { key: 'value' } });
   *
   * // Set default.
   * Container.setDefault(MyClass, MyClassFactory);
   *
   * // With default.
   * Container.get(MyClass);
   * ```
   * @param type The type or name of the class for which the default factory is set.
   * @param factory The factory class being set as default.
   */
  static setDefault<T>(type: Constructable<T> | string, factory: Factory<T>): void {
    const name = typeof type === 'string' ? type : type.name;

    this.factories[name]?.forEach((f) => {
      f.default = (f.factory as unknown as Constructable<T>).name === (factory as unknown as Constructable<T>).name;
    });
  }

  /**
   * The Container, by default, will wait a maximum of 5 seconds for a factory to resolve and provide an instance.
   * This method allows to change that timeout.
   *
   * @param timeoutInMs The timeout in milliseconds.
   */
  static setFactoryTimeout(timeoutInMs: number): void {
    this.FACTORY_TIMEOUT_IN_MS = timeoutInMs;
  }

  /**
   * `Container.unRegisterFactory()` will unregister all factories of a class.
   *
   * @param type The type or name of the class for which all factory is unregistered.
   */
  static unRegisterFactory<T>(type: Constructable<T> | string): void {
    const name = typeof type === 'string' ? type : type.name;
    delete this.factories[name];
  }
}
