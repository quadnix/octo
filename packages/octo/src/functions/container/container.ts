import type { Constructable } from '../../app.type.js';

type Factory<T> = { create: (...args: unknown[]) => Promise<T> };
type FactoryValue<T> = Factory<T> | [Promise<Factory<T>>, (factory: Factory<T>) => void];

/**
 * The Container class is a box to hold all registered factories.
 *
 * In Octo, we minimize the use of the `new` keyword and get instance of a class using its factory.
 * The factory has full control over how the instance is created.
 *
 * By using the Container and Factory concepts, it is possible to override internal class definitions at runtime.
 * It is an incredibly powerful tool to customize implementation.
 */
export class Container {
  private static FACTORY_TIMEOUT_IN_MS = 5000;

  private static readonly factories: { [key: string]: { factory: FactoryValue<any> } } = {};

  /**
   * `Container.get()` allows to get an instance of a class using its factory.
   *
   * @param type The type or name of the class to get.
   * @param options Allows selection of a specific factory to use to get the instance.
   * - The `args` option supplies arguments to the factory.
   * @returns The instance of the class.
   */
  static async get<T>(
    type: Constructable<T> | string,
    options?: {
      args?: unknown[];
    },
  ): Promise<T> {
    const args = options?.args || [];
    const name = typeof type === 'string' ? type : type.name;

    const factoryContainer: (typeof Container.factories)[keyof typeof Container.factories] | undefined =
      this.factories[name];
    if (factoryContainer?.factory) {
      if (!Array.isArray(factoryContainer.factory)) {
        return (factoryContainer.factory as Factory<T>).create(...args);
      } else {
        const factory = await (factoryContainer.factory[0] as Promise<Factory<T>>);
        return factory.create(...args);
      }
    }

    const newFactoryContainer = {} as (typeof Container.factories)[keyof typeof Container.factories];
    this.factories[name] = newFactoryContainer;

    let promiseResolver;
    let promiseTimeout;
    const promise = new Promise<Factory<T>>((resolve, reject) => {
      promiseTimeout = setTimeout(() => {
        reject(new Error(`Timed out waiting for factory ${name} to resolve!`));
      }, this.FACTORY_TIMEOUT_IN_MS);
      promiseResolver = resolve;
    });
    newFactoryContainer.factory = [promise, promiseResolver];

    const factory = await promise;
    if (promiseTimeout) {
      clearTimeout(promiseTimeout);
    }
    return factory.create(...args);
  }

  /**
   * `Container.registerFactory()` allows to register a factory for a class.
   *
   * @param type The type or name of the class for which the factory is registered.
   * @param factory The factory class being registered.
   */
  static registerFactory<T>(type: Constructable<T> | string, factory: Factory<T>): void {
    const name = typeof type === 'string' ? type : type.name;

    const factoryContainer = this.factories[name];
    if (factoryContainer?.factory) {
      // If factory is not a promise set by get() above, it has already been registered.
      if (!Array.isArray(factoryContainer.factory)) {
        return;
      }

      const resolve = factoryContainer.factory[1] as (factory: Factory<T>) => void;
      resolve(factory);
      factoryContainer.factory = factory;

      return;
    } else {
      this.factories[name] = { factory };
    }
  }

  /**
   * `Container.reset()` clears all registered factories and empties the container. This is mostly used in testing.
   */
  static reset(): void {
    this.FACTORY_TIMEOUT_IN_MS = 5000;

    for (const name in this.factories) {
      delete this.factories[name];
    }
  }

  /**
   * The Container, by default, will wait a maximum of 5 seconds for a factory to resolve and provide an instance.
   * This method allows to change that timeout.
   *
   * @param timeoutInMs The timeout in milliseconds.
   */
  static setFactoryTimeout(timeoutInMs: number): void {
    this.FACTORY_TIMEOUT_IN_MS = timeoutInMs;
  }

  /**
   * `Container.unRegisterFactory()` will unregister all factories of a class.
   *
   * @param type The type or name of the class for which all factory is unregistered.
   */
  static unRegisterFactory<T>(type: Constructable<T> | string): void {
    const name = typeof type === 'string' ? type : type.name;
    delete this.factories[name];
  }

  static async waitToResolveAllFactories(): Promise<void> {
    const promiseToResolveAllFactories: Promise<Factory<unknown>>[] = [];

    for (const name of Object.keys(this.factories)) {
      const factoryContainer = this.factories[name];
      if (factoryContainer?.factory && Array.isArray(factoryContainer.factory)) {
        promiseToResolveAllFactories.push(factoryContainer.factory[0]);
      }
    }

    await Promise.all(promiseToResolveAllFactories);
  }
}
