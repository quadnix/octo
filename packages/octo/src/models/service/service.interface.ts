import { Service } from './service.model.js';

export interface IService {
  serviceId: Service['serviceId'];
}
