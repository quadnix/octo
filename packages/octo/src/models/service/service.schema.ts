import { Validate } from '../../decorators/validate.decorator.js';
import { Schema } from '../../functions/schema/schema.js';

export class ServiceSchema {
  /**
   * The ID of the service.
   */
  @Validate({ options: { maxLength: 128, minLength: 2, regex: /^[a-zA-Z0-9][\w.-]*[a-zA-Z0-9]$/ } })
  serviceId = Schema<string>();
}
