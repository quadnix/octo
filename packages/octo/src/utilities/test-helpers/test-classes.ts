import { NodeType, type UnknownAnchor, type UnknownModel, type UnknownResource } from '../../app.type.js';
import {
  AAnchor,
  AModule,
  AOverlay,
  AResource,
  ASharedResource,
  App,
  type BaseAnchorSchema,
  type BaseOverlaySchema,
  type BaseResourceSchema,
  type Diff,
  Schema,
} from '../../index.js';
import { AModel } from '../../models/model.abstract.js';

export class SharedTestResource extends ASharedResource<BaseResourceSchema, TestResource> {
  static override readonly NODE_NAME: string = 'test-resource';
  static override readonly NODE_PACKAGE: string = '@octo';
  static override readonly NODE_SCHEMA = {};
  static override readonly NODE_TYPE: NodeType = NodeType.SHARED_RESOURCE;

  constructor(resourceId: string, properties: { [key: string]: unknown }, parents: TestResource[]) {
    super(resourceId, properties, parents as AResource<BaseResourceSchema, TestResource>[]);
  }
}

export class TestAction {}

export class TestAnchor extends AAnchor<BaseAnchorSchema, UnknownModel> {
  static override readonly NODE_PACKAGE: string = '@octo';

  constructor(anchorId: string, properties: BaseAnchorSchema['properties'], parent: UnknownModel) {
    super(anchorId, properties, parent);
  }
}

export class TestAppModuleSchema {
  name = Schema<string>();
}
export class TestAppModule extends AModule<TestAppModuleSchema, App> {
  static override readonly MODULE_PACKAGE = '@octo';

  static override readonly MODULE_SCHEMA = TestAppModuleSchema;

  async onInit(inputs: TestAppModuleSchema): Promise<App> {
    return new App(inputs.name);
  }
}

export class TestModelWithoutUnsynth extends AModel<object, TestModelWithoutUnsynth> {
  static override readonly NODE_NAME: string = 'test-model';
  static override readonly NODE_PACKAGE: string = '@octo';
  static override readonly NODE_SCHEMA = {};
  static override readonly NODE_TYPE: NodeType = NodeType.MODEL;

  override async diffProperties(): Promise<Diff[]> {
    return [];
  }

  override setContext(): string {
    return 'test-model=test';
  }

  override synth(): object {
    return {};
  }
}

export class TestOverlay extends AOverlay<BaseOverlaySchema, TestOverlay> {
  static override readonly NODE_NAME: string = 'test-overlay';
  static override readonly NODE_PACKAGE: string = '@octo';
  static override readonly NODE_SCHEMA = {};
  static override readonly NODE_TYPE: NodeType = NodeType.OVERLAY;

  constructor(
    overlayId: BaseOverlaySchema['overlayId'],
    properties: BaseOverlaySchema['properties'],
    anchors: UnknownAnchor[],
  ) {
    super(overlayId, properties, anchors);
  }
}

export class TestOverlayModuleSchema {
  anchorName = Schema<string>();
  app = Schema<App>();
}
export class TestOverlayModule extends AModule<TestOverlayModuleSchema, TestOverlay> {
  static override readonly MODULE_PACKAGE = '@octo';

  static override readonly MODULE_SCHEMA = TestOverlayModuleSchema;

  async onInit(inputs: TestOverlayModuleSchema): Promise<TestOverlay> {
    const anchor = inputs.app.getAnchor(inputs.anchorName)!;
    return new TestOverlay('test-overlay', {}, [anchor]);
  }
}

export class TestResource extends AResource<BaseResourceSchema, TestResource> {
  static override readonly NODE_NAME: string = 'test-resource';
  static override readonly NODE_PACKAGE: string = '@octo';
  static override readonly NODE_SCHEMA = {};
  static override readonly NODE_TYPE: NodeType = NodeType.RESOURCE;

  constructor(resourceId: string, properties: BaseResourceSchema['properties'] = {}, parents: UnknownResource[] = []) {
    super(resourceId, properties, parents);
  }
}

export class TestResourceWithDiffOverride extends AResource<BaseResourceSchema, TestResource> {
  static override readonly NODE_NAME: string = 'test-resource';
  static override readonly NODE_PACKAGE: string = '@octo';
  static override readonly NODE_SCHEMA = {};
  static override readonly NODE_TYPE: NodeType = NodeType.RESOURCE;

  constructor(resourceId: string, properties: BaseResourceSchema['properties'] = {}, parents: UnknownResource[] = []) {
    super(resourceId, properties, parents);
  }

  override async diff(): Promise<Diff[]> {
    return [];
  }
}
