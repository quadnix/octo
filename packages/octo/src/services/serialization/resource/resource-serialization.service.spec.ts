import { SharedTestResource, TestResource } from '../../../../test/helpers/test-classes.js';
import { type UnknownResource } from '../../../app.type.js';
import { Container } from '../../../decorators/container.js';
import { ResourceDataRepository, ResourceDataRepositoryFactory } from '../../../resources/resource-data.repository.js';
import { type AResource } from '../../../resources/resource.abstract.js';
import { ResourceSerializationService, ResourceSerializationServiceFactory } from './resource-serialization.service.js';

describe('Resource Serialization Service UT', () => {
  beforeEach(() => {
    Container.registerFactory(ResourceDataRepository, ResourceDataRepositoryFactory);
    Container.get(ResourceDataRepository, { args: [true, [], []] });

    Container.registerFactory(ResourceSerializationService, ResourceSerializationServiceFactory);
    Container.get(ResourceSerializationService, { args: [true] });
  });

  afterEach(() => {
    Container.reset();
  });

  describe('deserialize()', () => {
    it('should throw error when de-serializing an unknown class', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';

      const resources = [resource1];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });

      await expect(async () => {
        const serializedOutput = await service.serialize();
        await service.deserialize(serializedOutput);
      }).rejects.toThrowErrorMatchingInlineSnapshot(`"Cannot read properties of undefined (reading 'unSynth')"`);
    });

    it('should deserialize a single resource', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';

      const resources = [resource1];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });
      service.registerClass('TestResource', TestResource);

      const serializedOutput = await service.serialize();
      await service.deserialize(serializedOutput);

      const resourceDataRepository = await Container.get(ResourceDataRepository);
      const deserializedResources = resourceDataRepository['oldResources'];
      expect(deserializedResources).toMatchSnapshot();
    });

    it('should deserialize a resource with complex properties', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = { key2: { key3: 'value3' }, key4: 'value4' };
      resource1.response['response1'] = 'value1';

      const resources = [resource1];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService);
      service.registerClass('TestResource', TestResource);

      const serializedOutput = await service.serialize();
      await service.deserialize(serializedOutput);

      const resourceDataRepository = await Container.get(ResourceDataRepository);
      const deserializedResources = resourceDataRepository['oldResources'];
      expect(deserializedResources).toMatchSnapshot();
    });

    it('should deserialize a single shared resource', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';
      const resource2 = new SharedTestResource('resource-2', {}, [resource1]);

      const resources = [resource1, resource2];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService);
      service.registerClass('TestResource', TestResource);
      service.registerClass('SharedTestResource', SharedTestResource);

      const serializedOutput = await service.serialize();
      await service.deserialize(serializedOutput);

      const resourceDataRepository = await Container.get(ResourceDataRepository);
      const deserializedResources = resourceDataRepository['oldResources'];
      expect(deserializedResources).toMatchSnapshot();
    });

    it('should throw error trying to deserialize a shared resource without corresponding resource', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';
      const resource2 = new SharedTestResource('resource-2', {}, [resource1]);

      const resources = [resource2];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });
      service.registerClass('TestResource', TestResource);
      service.registerClass('SharedTestResource', SharedTestResource);
      service.setResourceDeserializationTimeout(50);

      const serializedOutput = await service.serialize();

      await expect(async () => {
        await service.deserialize(serializedOutput);
      }).rejects.toMatchInlineSnapshot(`[Error: DeReferencing resource operation timed out!]`);
    });

    it('should deserialize dependencies', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';
      const resource2 = new TestResource('resource-2');
      resource2.properties['key2'] = 'value2';
      resource2.response['response2'] = 'value2';
      resource1.addChild('resourceId', resource2, 'resourceId');

      const resources = [resource1, resource2];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService);
      service.registerClass('TestResource', TestResource);

      const serializedOutput = await service.serialize();
      await service.deserialize(serializedOutput);

      const resourceDataRepository = await Container.get(ResourceDataRepository);
      const deserializedResources = resourceDataRepository['oldResources'];

      const resource1Deserialized = deserializedResources.find((r) => r.resourceId === 'resource-1');
      const resource2Deserialized = resource1Deserialized!.getChildren()['test-resource'][0]
        .to as AResource<TestResource>;
      expect(resource2Deserialized.resourceId).toBe('resource-2');
      expect((resource2Deserialized.getParents()['test-resource'][0].to as AResource<TestResource>).resourceId).toBe(
        'resource-1',
      );
    });

    it('should deserialize dependencies in reverse order', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';
      const resource2 = new TestResource('resource-2');
      resource2.properties['key2'] = 'value2';
      resource2.response['response2'] = 'value2';
      resource1.addChild('resourceId', resource2, 'resourceId');

      const resources = [resource2, resource1]; // Reverse order of resources.
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService);
      service.registerClass('TestResource', TestResource);

      const serializedOutput = await service.serialize();
      await service.deserialize(serializedOutput);

      const resourceDataRepository = await Container.get(ResourceDataRepository);
      const deserializedResources = resourceDataRepository['oldResources'];

      const resource1Deserialized = deserializedResources.find((r) => r.resourceId === 'resource-1');
      const resource2Deserialized = resource1Deserialized!.getChildren()['test-resource'][0]
        .to as AResource<TestResource>;
      expect(resource2Deserialized.resourceId).toBe('resource-2');
      expect((resource2Deserialized.getParents()['test-resource'][0].to as AResource<TestResource>).resourceId).toBe(
        'resource-1',
      );
    });

    it('should have exact same dependencies on deserialized object as original object', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';
      const resource2 = new TestResource('resource-2');
      resource2.properties['key2'] = 'value2';
      resource2.response['response2'] = 'value2';
      resource1.addChild('resourceId', resource2, 'resourceId');
      const resource3 = new SharedTestResource('resource-3', {}, [resource1]);

      const resources = [resource1, resource2, resource3];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService);
      service.registerClass('TestResource', TestResource);
      service.registerClass('SharedTestResource', SharedTestResource);

      const serializedOutput1 = await service.serialize();
      const currentResourceDependencies = serializedOutput1.dependencies.sort((a, b) =>
        a.from + a.to > b.from + b.to ? 1 : b.from + b.to > a.from + a.to ? -1 : 0,
      );

      await service.deserialize(serializedOutput1);
      const resourceDataRepository = await Container.get(ResourceDataRepository);
      const deserializedResources = resourceDataRepository['oldResources'];

      // An easy way to get dependencies from `deserializedResources` is to serialize it again.
      await Container.get(ResourceDataRepository, {
        args: [true, [...deserializedResources], [...deserializedResources]],
      });
      const serializedOutput2 = await service.serialize();
      const previousResourcesDependencies = serializedOutput2.dependencies.sort((a, b) =>
        a.from + a.to > b.from + b.to ? 1 : b.from + b.to > a.from + a.to ? -1 : 0,
      );

      expect(previousResourcesDependencies).toEqual(currentResourceDependencies);
    });

    it('should initialize ResourceDataRepository with separate old and new resources', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';

      const resources = [resource1];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });
      service.registerClass('TestResource', TestResource);

      const serializedOutput = await service.serialize();
      await service.deserialize(serializedOutput);

      const resourceDataRepository = await Container.get(ResourceDataRepository);

      // Manipulate new resources. This must not affect old resources in any way.
      resourceDataRepository.getById('resource-1')!.properties['key1'] = 'value2';

      expect(resourceDataRepository['newResources'][0]).not.toEqual(resourceDataRepository['oldResources'][0]);
    });

    it('should not initialize ResourceDataRepository with new resources marked for deletion', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';

      const resources = [resource1];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });
      service.registerClass('TestResource', TestResource);

      const serializedOutput1 = await service.serialize();
      await service.deserialize(serializedOutput1);

      const resourceDataRepository = await Container.get(ResourceDataRepository);
      expect(resourceDataRepository.getById('resource-1')).not.toBeUndefined();

      // Remove the new resource.
      resourceDataRepository.remove(resource1);

      // Ensure the deleted resources are not initialized in the ResourceDataRepository.
      expect(resourceDataRepository.getById('resource-1')).toBeUndefined();

      const serializedOutput2 = await service.serialize();
      await service.deserialize(serializedOutput2);

      // Ensure the deleted resources are not initialized in the ResourceDataRepository.
      expect(resourceDataRepository['oldResources'].length).toBe(0);
    });
  });

  describe('serialize()', () => {
    it('should serialize an empty array', async () => {
      await Container.get(ResourceDataRepository, { args: [true, [], []] });

      const service = await Container.get(ResourceSerializationService);
      expect(await service.serialize()).toMatchSnapshot();
    });

    it('should serialize non-empty array', async () => {
      const resources: UnknownResource[] = [new TestResource('resource-1')];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });
      expect(await service.serialize()).toMatchSnapshot();
    });

    it('should not serialize deleted resources', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';
      const resource2 = new TestResource('resource-2');
      resource2.properties['key2'] = 'value2';
      resource2.response['response2'] = 'value2';
      resource2.remove();

      const resources = [resource1, resource2];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });
      expect(await service.serialize()).toMatchSnapshot();
    });

    it('should serialize dependencies and properties and resources', async () => {
      const resource1 = new TestResource('resource-1');
      resource1.properties['key1'] = 'value1';
      resource1.response['response1'] = 'value1';
      const resource2 = new TestResource('resource-2');
      resource2.properties['key2'] = 'value2';
      resource2.response['response2'] = 'value2';
      resource1.addChild('resourceId', resource2, 'resourceId');

      const resources = [resource1, resource2];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });
      expect(await service.serialize()).toMatchSnapshot();
    });

    it('should serialize shared resources', async () => {
      const resource1 = new TestResource('resource-1');
      const resource2 = new SharedTestResource('resource-2', {}, [resource1]);
      resource2.properties['key2'] = 'value2';
      resource2.response['response2'] = 'value2';

      const resources = [resource1, resource2];
      await Container.get(ResourceDataRepository, { args: [true, [...resources], [...resources]] });

      const service = await Container.get(ResourceSerializationService, { args: [true] });
      expect(await service.serialize()).toMatchSnapshot();
    });
  });
});
