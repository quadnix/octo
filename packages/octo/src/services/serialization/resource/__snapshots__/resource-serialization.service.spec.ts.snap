// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource Serialization Service UT deserialize() should deserialize a resource with complex properties 1`] = `
[
  TestResource {
    "MODEL_NAME": "test-resource",
    "MODEL_TYPE": "resource",
    "_deleteMarker": false,
    "anchors": [],
    "context": undefined,
    "dependencies": [],
    "properties": {
      "key1": {
        "key2": {
          "key3": "value3",
        },
        "key4": "value4",
      },
    },
    "resourceId": "resource-1",
    "response": {
      "response1": "value1",
    },
  },
]
`;

exports[`Resource Serialization Service UT deserialize() should deserialize a single resource 1`] = `
[
  TestResource {
    "MODEL_NAME": "test-resource",
    "MODEL_TYPE": "resource",
    "_deleteMarker": false,
    "anchors": [],
    "context": undefined,
    "dependencies": [],
    "properties": {
      "key1": "value1",
    },
    "resourceId": "resource-1",
    "response": {
      "response1": "value1",
    },
  },
]
`;

exports[`Resource Serialization Service UT deserialize() should deserialize a single shared resource 1`] = `
[
  TestResource {
    "MODEL_NAME": "test-resource",
    "MODEL_TYPE": "resource",
    "_deleteMarker": false,
    "anchors": [],
    "context": undefined,
    "dependencies": [
      {
        "from": "test-resource=resource-1",
        "relationship": {
          "onField": "resourceId",
          "toField": "resourceId",
          "type": "parent",
        },
        "to": "test-resource=resource-2",
      },
    ],
    "properties": {
      "key1": "value1",
    },
    "resourceId": "resource-1",
    "response": {
      "response1": "value1",
    },
  },
  SharedTestResource {
    "MODEL_NAME": "test-resource",
    "MODEL_TYPE": "shared-resource",
    "_deleteMarker": false,
    "anchors": [],
    "context": "test-resource=resource-2",
    "dependencies": [
      {
        "from": "test-resource=resource-2",
        "relationship": {
          "onField": "resourceId",
          "toField": "resourceId",
          "type": "child",
        },
        "to": "test-resource=resource-1",
      },
    ],
    "properties": {},
    "resourceId": "resource-2",
    "response": {},
  },
]
`;

exports[`Resource Serialization Service UT serialize() should not serialize deleted resources 1`] = `
{
  "dependencies": [],
  "resources": {
    "resource-1": {
      "className": "TestResource",
      "resource": {
        "properties": {
          "key1": "value1",
        },
        "resourceId": "resource-1",
        "response": {
          "response1": "value1",
        },
      },
    },
  },
  "sharedResources": {},
}
`;

exports[`Resource Serialization Service UT serialize() should serialize an empty array 1`] = `
{
  "dependencies": [],
  "resources": {},
  "sharedResources": {},
}
`;

exports[`Resource Serialization Service UT serialize() should serialize dependencies and properties and resources 1`] = `
{
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "test-resource=resource-2",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-2",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "test-resource=resource-1",
    },
  ],
  "resources": {
    "resource-1": {
      "className": "TestResource",
      "resource": {
        "properties": {
          "key1": "value1",
        },
        "resourceId": "resource-1",
        "response": {
          "response1": "value1",
        },
      },
    },
    "resource-2": {
      "className": "TestResource",
      "resource": {
        "properties": {
          "key2": "value2",
        },
        "resourceId": "resource-2",
        "response": {
          "response2": "value2",
        },
      },
    },
  },
  "sharedResources": {},
}
`;

exports[`Resource Serialization Service UT serialize() should serialize non-empty array 1`] = `
{
  "dependencies": [],
  "resources": {
    "resource-1": {
      "className": "TestResource",
      "resource": {
        "properties": {},
        "resourceId": "resource-1",
        "response": {},
      },
    },
  },
  "sharedResources": {},
}
`;

exports[`Resource Serialization Service UT serialize() should serialize shared resources 1`] = `
{
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "test-resource=resource-2",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-2",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "test-resource=resource-1",
    },
  ],
  "resources": {
    "resource-1": {
      "className": "TestResource",
      "resource": {
        "properties": {},
        "resourceId": "resource-1",
        "response": {},
      },
    },
  },
  "sharedResources": {
    "resource-2": {
      "className": "SharedTestResource",
      "resource": {
        "properties": {
          "key2": "value2",
        },
        "resourceId": "resource-2",
        "response": {
          "response2": "value2",
        },
      },
    },
  },
}
`;
